lib/seahorn/BvOpSem2TrackingRawMemMgr.hh

```diff
@@ -40,23 +45,53 @@ public:
   struct MemValTyImpl {
     Expr m_v;
 
-    MemValTyImpl(RawMemValTy &&raw_val, Expr &&metadata_val) {
+    MemValTyImpl(RawMemValTy &&raw_val, RawMemValTy &&r_metadata_val,
+                 RawMemValTy &&w_metadata_val, RawMemValTy &&a_metadata_val) {
       assert(!strct::isStructVal(raw_val));
-      assert(!strct::isStructVal(metadata_val));
-      m_v = strct::mk(std::move(raw_val), std::move(metadata_val));
+      assert(!strct::isStructVal(r_metadata_val));
+      assert(!strct::isStructVal(w_metadata_val));
+      assert(!strct::isStructVal(a_metadata_val));
+      llvm::SmallVector<RawMemValTy, 4> kids;
+      kids.push_back(std::move(raw_val));
+      kids.push_back(std::move(r_metadata_val));
+      kids.push_back(std::move(w_metadata_val));
+      kids.push_back(std::move(a_metadata_val));
+
+      m_v = strct::mk(kids);
     }
 
-    MemValTyImpl(const RawMemValTy &raw_val, const RawMemValTy &metadata_val) {
+    MemValTyImpl(const RawMemValTy &raw_val, const RawMemValTy &r_metadata_val,
+                 const RawMemValTy &w_metadata_val,
+                 const RawMemValTy &a_metadata_val) {
       assert(!strct::isStructVal(raw_val));
-      assert(!strct::isStructVal(metadata_val));
-      m_v = strct::mk(raw_val, metadata_val);
+      assert(!strct::isStructVal(r_metadata_val));
+      assert(!strct::isStructVal(w_metadata_val));
+      assert(!strct::isStructVal(a_metadata_val));
+      llvm::SmallVector<RawMemValTy, 4> kids;
+      kids.push_back(std::move(raw_val));
+      kids.push_back(std::move(r_metadata_val));
+      kids.push_back(std::move(w_metadata_val));
+      kids.push_back(std::move(a_metadata_val));
+
+      m_v = strct::mk(kids);
+    }
+
+    MemValTyImpl(MetadataKind kind, const MemValTyImpl &orig_val,
+                 const RawMemValTy &raw_val) {
+      llvm::SmallVector<RawMemValTy, 4> kids;
+      for (auto i = 0; i < orig_val.toExpr()->arity(); i++) {
+        kids.push_back(i == kind + 1 ? raw_val : orig_val.toExpr()->arg(i));
```

Comment:
added comment.

Reply:
also comment what is `i == kind + 1` -- this depends on internal representation of meta-data fields.
update `for` loop per my comment -- only call `arity()` ones, use pre-inc instead of post-inc. These are my nits, but they are not hard to change...